src/workerd/jsg/memory.h

```diff
@@ -0,0 +1,810 @@
+    trackField(nullptr, *it, elementName);
+  }
+  popNode();
+  return *this;
+}
+
+template <MemoryRetainer T, typename Iterator>
+MemoryTracker& MemoryTracker::trackField(
+    kj::StringPtr edgeName,
+    const T& value,
+    kj::Maybe<kj::StringPtr> nodeName,
+    kj::Maybe<kj::StringPtr> elementName,
+    bool subtractFromSelf) {
+  if (value.begin() == value.end()) return *this;
+  KJ_IF_SOME(currentNode, getCurrentNode()) {
+    if (subtractFromSelf) {
+      currentNode.size_ -= sizeof(T);
+    }
+  }
+  pushNode(getNodeName(nodeName, edgeName), sizeof(T), edgeName);
+  for (Iterator it = value.begin(); it != value.end(); ++it) {
+    // Use nullptr as edge names so the elements appear as indexed properties
+    trackField(nullptr, *it, elementName);
+  }
+  popNode();
+  return *this;
+}
+
+template <MemoryRetainer T>
+MemoryTracker& MemoryTracker::trackField(
+    kj::StringPtr edgeName,
+    const kj::ArrayPtr<T>& value,
+    kj::Maybe<kj::StringPtr> nodeName,
+    kj::Maybe<kj::StringPtr> elementName,
+    bool subtractFromSelf) {
+  if (value.begin() == value.end()) return *this;
+  KJ_IF_SOME(currentNode, getCurrentNode()) {
+    if (subtractFromSelf) {
+      currentNode.size_ -= sizeof(T);
+    }
+  }
+  pushNode(getNodeName(nodeName, edgeName), sizeof(T), edgeName);
+  for (const auto item : value) {
+    // Use nullptr as edge names so the elements appear as indexed properties
+    trackField(nullptr, item, elementName);
+  }
+  popNode();
+  return *this;
+}
+
+template <MemoryRetainer T>
+MemoryTracker& MemoryTracker::trackField(
+    kj::StringPtr edgeName,
+    const kj::ArrayPtr<T* const>& value,
+    kj::Maybe<kj::StringPtr> nodeName,
+    kj::Maybe<kj::StringPtr> elementName,
+    bool subtractFromSelf) {
+  if (value.begin() == value.end()) return *this;
+  KJ_IF_SOME(currentNode, getCurrentNode()) {
+    if (subtractFromSelf) {
+      currentNode.size_ -= sizeof(T);
+    }
+  }
+  pushNode(getNodeName(nodeName, edgeName), sizeof(T), edgeName);
+  for (const auto item : value) {
+    // Use nullptr as edge names so the elements appear as indexed properties
+    trackField(nullptr, item, elementName);
+  }
+  popNode();
+  return *this;
+}
+
+template <MemoryRetainer T>
+MemoryTracker& MemoryTracker::trackField(
+    kj::StringPtr edgeName,
+    const T& value,
+    kj::Maybe<kj::StringPtr> nodeName) {
+  return trackField(edgeName, &value, nodeName);
+}
+
+template <MemoryRetainer T>
+MemoryTracker& MemoryTracker::trackField(
+    kj::StringPtr edgeName,
+    const T* value,
+    kj::Maybe<kj::StringPtr> nodeName) {
+  if (value == nullptr) return *this;
+  KJ_IF_SOME(found, seen_.find(value)) {
+    KJ_IF_SOME(currentNode, getCurrentNode()) {
+      graph_->AddEdge(&currentNode, found.node, edgeName.cStr());
+    } else {
+      graph_->AddEdge(nullptr, found.node, edgeName.cStr());
+    }
+    return *this;
+  }
+
+  return track(value, edgeName);
+}
+
+template <typename T, typename test_for_number, typename dummy>
```

Comment:
Also unused?

```suggestion
template <typename T>
```

Reply:
It is used. The additional template parameters create a specialization for when `T` is numeric. Will keep these as is for now.